File: backend/routes/payroll.py
from flask import Blueprint, jsonify, request
from database import get_db
from datetime import datetime, date
import calendar

payroll_bp = Blueprint('payroll', __name__)

TAX_BRACKETS = [
    (235000,    0),
    (335000,  0.10),
    (410000,  0.20),
    (10000000,0.30),
    (float('inf'), 0.40)
]

def calc_tax(gross):
    tax = 0
    prev = 0
    brackets = [(235000,0),(335000,.10),(410000,.20),(10000000,.30),(float('inf'),.40)]
    limits = [235000, 335000, 410000, 10000000]
    for i,(limit,rate) in enumerate(brackets):
        if gross <= prev: break
        taxable = min(gross, limit) - prev
        tax += taxable * rate
        prev = limit
    return round(tax, 2)

@payroll_bp.route('/', methods=['GET'])
def list_payroll():
    db = get_db()
    period = request.args.get('period', '')
    dept = request.args.get('department_id', '')
    status = request.args.get('status', '')
    sql = """
        SELECT p.*, e.first_name||' '||e.last_name AS name, e.employee_no,
               d.name AS department
        FROM payroll p
        JOIN employees e ON e.id = p.employee_id
        LEFT JOIN departments d ON d.id = e.department_id
        WHERE 1=1
    """
    params = []
    if period:
        sql += " AND strftime('%Y-%m', p.period_start) = ?"
        params.append(period)
    if dept:
        sql += " AND e.department_id = ?"
        params.append(dept)
    if status:
        sql += " AND p.status = ?"
        params.append(status)
    sql += " ORDER BY p.period_start DESC, e.first_name"
    rows = db.execute(sql, params).fetchall()
    db.close()
    return jsonify([dict(r) for r in rows])

@payroll_bp.route('/process', methods=['POST'])
def process_payroll():
    """Process payroll for a period for all or specific employees."""
    db = get_db()
    data = request.json
    period_start = data.get('period_start')
    period_end = data.get('period_end')
    dept_id = data.get('department_id')

    if not period_start or not period_end:
        return jsonify({'error': 'period_start and period_end required'}), 400

    sql = "SELECT * FROM employees WHERE status='active'"
    params = []
    if dept_id:
        sql += " AND department_id=?"
        params.append(dept_id)
    employees = db.execute(sql, params).fetchall()

    # Count working days in period
    from datetime import timedelta
    ps = datetime.strptime(period_start, '%Y-%m-%d').date()
    pe = datetime.strptime(period_end, '%Y-%m-%d').date()
    work_days = sum(1 for d in range((pe-ps).days+1) if (ps+timedelta(d)).weekday()<5)

    processed = []
    for emp in employees:
        # Actual attendance in period
        att = db.execute("""
            SELECT COUNT(*) as days_worked, SUM(overtime_hrs) as ot
            FROM attendance
            WHERE employee_id=? AND date BETWEEN ? AND ? AND status='present'
        """, (emp['id'], period_start, period_end)).fetchone()

        days_worked = att['days_worked'] or 0
        days_absent = work_days - days_worked
        overtime_hrs = att['ot'] or 0

        daily_rate = emp['salary'] / 22  # ~22 working days/month
        base = round(emp['salary'] * (days_worked / work_days), 2) if work_days else emp['salary']
        allowances = round(emp['salary'] * 0.15, 2)  # 15% allowances
        overtime_pay = round(daily_rate / 8 * 1.5 * overtime_hrs, 2)
        gross = base + allowances + overtime_pay
        tax = calc_tax(gross)
        nssf = round(gross * 0.05, 2)  # 5% NSSF employee contribution
        deductions = round(nssf, 2)
        net_pay = round(gross - tax - deductions, 2)

        # Remove existing draft for this period
        db.execute("DELETE FROM payroll WHERE employee_id=? AND period_start=? AND status='draft'",
                   (emp['id'], period_start))
        db.execute("""
            INSERT INTO payroll
            (employee_id,period_start,period_end,base_salary,allowances,deductions,
             overtime_pay,tax,net_pay,days_worked,days_absent,status,processed_at)
            VALUES (?,?,?,?,?,?,?,?,?,?,?,'processed',datetime('now'))
        """, (emp['id'], period_start, period_end, base, allowances, deductions,
              overtime_pay, tax, net_pay, days_worked, days_absent))
        processed.append({'employee_id': emp['id'], 'net_pay': net_pay})

    db.commit(); db.close()
    total_payout = sum(p['net_pay'] for p in processed)
    return jsonify({
        'message': f'Payroll processed for {len(processed)} employees',
        'count': len(processed),
        'total_payout': total_payout,
        'period': f'{period_start} — {period_end}'
    })

@payroll_bp.route('/summary', methods=['GET'])
def payroll_summary():
    db = get_db()
    period = request.args.get('period', datetime.now().strftime('%Y-%m'))
    rows = db.execute("""
        SELECT d.name AS department, SUM(p.net_pay) AS total_net,
               SUM(p.tax) AS total_tax, SUM(p.base_salary) AS total_base,
               COUNT(*) AS employee_count
        FROM payroll p
        JOIN employees e ON e.id = p.employee_id
        LEFT JOIN departments d ON d.id = e.department_id
        WHERE strftime('%Y-%m', p.period_start) = ? AND p.status='processed'
        GROUP BY d.id
    """, (period,)).fetchall()
    totals = db.execute("""
        SELECT SUM(net_pay) as net, SUM(tax) as tax, SUM(base_salary) as base,
               SUM(overtime_pay) as ot, COUNT(*) as count
        FROM payroll WHERE strftime('%Y-%m', period_start) = ? AND status='processed'
    """, (period,)).fetchone()
    db.close()
    return jsonify({
        'period': period,
        'by_department': [dict(r) for r in rows],
        'totals': dict(totals) if totals else {}
    })

File: frontend/static/js/pages/payroll.js
// Payroll Page
async function renderPayroll() {
  const now = new Date();
  const month = now.toISOString().slice(0,7);
  const firstDay = month + '-01';
  const lastDay = new Date(now.getFullYear(), now.getMonth()+1, 0).toISOString().split('T')[0];

  document.getElementById('page-content').innerHTML = `
    <div class="page-topbar">
      <div class="page-title-block">
        <h1 class="page-title">Payroll</h1>
        <p class="page-subtitle">Process salaries, deductions, and tax computations</p>
      </div>
      <div class="page-actions">
        <button class="btn btn-primary" onclick="openProcessModal('${firstDay}','${lastDay}')">Process Payroll</button>
      </div>
    </div>
    <div class="page-body">
      <div id="payrollSummaryCards" class="stat-grid">
        ${[1,2,3,4].map(()=>`<div class="stat-card"><div class="skeleton" style="height:70px"></div></div>`).join('')}
      </div>

      <div class="toolbar">
        <div class="form-group" style="flex-direction:row;align-items:center;gap:8px;margin:0">
          <label class="form-label">PERIOD</label>
          <input type="month" class="form-control" id="payrollMonth" value="${month}" style="width:170px" onchange="loadPayrollTable()"/>
        </div>
        <select class="form-control" id="payrollStatus" style="width:140px" onchange="loadPayrollTable()">
          <option value="">All</option>
          <option value="processed">Processed</option>
          <option value="draft">Draft</option>
        </select>
        <div class="toolbar-spacer"></div>
        <span class="tag tag-muted" id="payrollCountTag">—</span>
      </div>

      <div class="card">
        <div class="table-scroll">
          <table class="data-table" role="grid" aria-label="Payroll records">
            <thead>
              <tr>
                <th>Employee</th>
                <th>Department</th>
                <th>Period</th>
                <th>Days Worked</th>
                <th>Base Salary</th>
                <th>Allowances</th>
                <th>Overtime</th>
                <th>Tax</th>
                <th>Net Pay</th>
                <th>Status</th>
              </tr>
            </thead>
            <tbody id="payrollTableBody">${loadingRows(10,5)}</tbody>
          </table>
        </div>
      </div>
    </div>`;

  loadPayrollSummary(month);
  loadPayrollTable();
}

async function loadPayrollSummary(period) {
  try {
    const s = await API.payroll.summary(period);
    const t = s.totals || {};
    document.getElementById('payrollSummaryCards').innerHTML = `
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--citron)"></div>
        <div class="stat-label">Gross Payroll</div>
        <div class="stat-value amount gross" style="font-size:24px">${fmtCurrency(t.base||0,'UGX')}</div>
        <div class="stat-delta">${t.count||0} employees processed</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--mint)"></div>
        <div class="stat-label">Net Payout</div>
        <div class="stat-value amount net" style="font-size:24px">${fmtCurrency(t.net||0,'UGX')}</div>
        <div class="stat-delta up">After all deductions</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--coral)"></div>
        <div class="stat-label">Total Tax</div>
        <div class="stat-value amount tax" style="font-size:24px">${fmtCurrency(t.tax||0,'UGX')}</div>
        <div class="stat-delta dn">PAYE withheld</div>
      </div>
      <div class="stat-card">
        <div class="stat-card-accent" style="background:var(--periwinkle)"></div>
        <div class="stat-label">Employees Paid</div>
        <div class="stat-value" style="color:var(--periwinkle)">${t.count||0}</div>
        <div class="stat-delta">This period</div>
      </div>`;
  } catch(e) {
    document.getElementById('payrollSummaryCards').innerHTML = `
      ${[['Gross Payroll','—'],['Net Payout','—'],['Total Tax','—'],['Employees Paid','0']].map(([l,v])=>`
        <div class="stat-card"><div class="stat-label">${l}</div><div class="stat-value" style="font-size:24px">${v}</div></div>`).join('')}`;
  }
}

async function loadPayrollTable() {
  const period = document.getElementById('payrollMonth')?.value || '';
  const status = document.getElementById('payrollStatus')?.value || '';
  try {
    const data = await API.payroll.list({ period, status });
    document.getElementById('payrollCountTag').textContent = `${data.length} records`;
    if(!data.length) {
      document.getElementById('payrollTableBody').innerHTML = `<tr><td colspan="10"><div class="empty-state"><div class="empty-icon">o</div>No payroll processed for this period.<br/><button class="btn btn-primary" style="margin-top:12px" onclick="openProcessModal()">Process Now</button></div></td></tr>`;
      return;
    }
    document.getElementById('payrollTableBody').innerHTML = data.map(r=>`
      <tr>
        <td>
          <div style="display:flex;align-items:center;gap:10px">
            ${avatar(r.name)}
            <div>
              <div style="font-weight:600">${r.name}</div>
              <div style="font-size:11px;color:var(--text-2);font-family:var(--font-mono)">${r.employee_no}</div>
            </div>
          </div>
        </td>
        <td style="color:var(--text-2)">${r.department||'—'}</td>
        <td style="font-family:var(--font-mono);font-size:11px;color:var(--text-2)">${r.period_start} — ${r.period_end}</td>
        <td style="text-align:center">
          <span style="font-family:var(--font-mono)">
            <span style="color:var(--mint)">${r.days_worked}</span>/<span style="color:var(--coral)">${r.days_absent}</span>
          </span>
        </td>
        <td class="amount">${fmtCurrency(r.base_salary,'UGX')}</td>
        <td class="amount" style="color:var(--amber)">${fmtCurrency(r.allowances,'UGX')}</td>
        <td class="amount" style="color:var(--periwinkle)">${fmtCurrency(r.overtime_pay,'UGX')}</td>
        <td class="amount tax">${fmtCurrency(r.tax,'UGX')}</td>
        <td class="amount net">${fmtCurrency(r.net_pay,'UGX')}</td>
        <td>${statusTag(r.status)}</td>
      </tr>`).join('');
    loadPayrollSummary(period);
  } catch(e) { toast('Failed: '+e.message,'error'); }
}

function openProcessModal(start='', end='') {
  const now = new Date();
  const defaultStart = start || new Date(now.getFullYear(),now.getMonth(),1).toISOString().split('T')[0];
  const defaultEnd   = end   || new Date(now.getFullYear(),now.getMonth()+1,0).toISOString().split('T')[0];
  openModal(`
    <div class="modal-header">
      <span class="modal-title">Process Payroll</span>
      <button class="modal-close" onclick="closeModal()">x</button>
    </div>
    <div class="modal-body">
      <div style="background:var(--obsidian3);border:1px solid var(--border);border-radius:8px;padding:14px;margin-bottom:14px;font-size:13px;color:var(--text-2)">
        This will compute salaries based on actual attendance data, apply PAYE tax brackets, NSSF deductions, and calculate overtime pay.
      </div>
      <div class="form-row">
        <div class="form-group"><label class="form-label">Period Start *</label><input class="form-control" id="procStart" type="date" value="${defaultStart}"/></div>
        <div class="form-group"><label class="form-label">Period End *</label><input class="form-control" id="procEnd" type="date" value="${defaultEnd}"/></div>
      </div>
      <div class="form-group">
        <label class="form-label">Department (optional — leave blank for all)</label>
        <select class="form-control" id="procDept"><option value="">All Departments</option></select>
      </div>
    </div>
    <div class="modal-footer">
      <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
      <button class="btn btn-primary" id="procBtn" onclick="submitProcess()">Process Payroll</button>
    </div>`);
  API.departments.list().then(depts => {
    const sel = document.getElementById('procDept');
    if(sel) sel.innerHTML += depts.map(d=>`<option value="${d.id}">${d.name}</option>`).join('');
  });
}

async function submitProcess() {
  const body = {
    period_start: document.getElementById('procStart').value,
    period_end:   document.getElementById('procEnd').value,
    department_id: document.getElementById('procDept').value || null,
  };
  if(!body.period_start||!body.period_end) { toast('Period required','error'); return; }
  const btn = document.getElementById('procBtn');
  btn.disabled=true; btn.textContent='Processing...';
  try {
    const r = await API.payroll.process(body);
    toast(`${r.message} — Total: ${fmtCurrency(r.total_payout,'UGX')}`,'success');
    closeModal(); loadPayrollTable();
  } catch(e) { toast('Error: '+e.message,'error'); btn.disabled=false; btn.textContent='Process Payroll'; }
}
